p2p/raintree/network.go

```diff
@@ -25,46 +25,71 @@ type rainTreeNetwork struct {
 
 	peersManager *peersManager
 
+	cleanupLayerEnabled    bool // debug config only
+	redundancyLayerEnabled bool
+
 	// TECHDEBT(drewsky): What should we use for de-duping messages within P2P?
 	mempool map[uint64]struct{} // TODO (drewsky) replace map implementation (can grow unbounded)
 }
 
-func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesP2P.AddrBook) typesP2P.Network {
+func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesP2P.AddrBook, cfg modules.P2PConfig) typesP2P.Network {
 	pm, err := newPeersManager(addr, addrBook)
 	if err != nil {
 		log.Println("[ERROR] Error initializing rainTreeNetwork peersManager: ", err)
 	}
 
 	n := &rainTreeNetwork{
-		selfAddr:     addr,
-		peersManager: pm,
-		mempool:      make(map[uint64]struct{}),
+		selfAddr:               addr,
+		peersManager:           pm,
+		redundancyLayerEnabled: cfg.GetUseRainTreeRedundancyLayer(),
+		cleanupLayerEnabled:    cfg.GetUseRainTreeCleanupLayer(),
+		mempool:                make(map[uint64]struct{}),
 	}
 
 	return typesP2P.Network(n)
 }
 
 func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
-	return n.networkBroadcastAtLevel(data, n.peersManager.getNetworkView().maxNumLevels, getNonce())
+	return n.networkBroadcastAtLevel(data, int32(n.peersManager.getNetworkView().maxNumLevels), getNonce())
 }
 
-func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level uint32, nonce uint64) error {
-	// This is handled either by the cleanup layer or redundancy layer
-	if level == 0 {
-		return nil
-	}
+func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level int32, nonce uint64) error {
 	msg := &typesP2P.RainTreeMessage{
 		Level: level,
 		Data:  data,
 		Nonce: nonce,
 	}
+
+	// Redundancy layer
+	if level == 0 {
+		if n.redundancyLayerEnabled {
+			level, msg.Level = n.redundancyLayer()
+		} else {
+			if err := n.demote(msg); err != nil {
+				log.Println("Error demoting self during RainTree message propagation: ", err)
+			}
+			return nil
+		}
```

Comment:
Same code structure as in the cleanup layer

```suggestion
		if !n.redundancyLayerEnabled {
			if err := n.demote(msg); err != nil {
				log.Println("Error demoting self during RainTree message propagation: ", err)
			}
			return nil
		}
		level, msg.Level = n.redundancyLayer()
```